chips/sam4l/src/pm.rs

```diff
@@ -254,44 +239,104 @@ pub enum PBDClock {
     PICOUART,
 }
 
+/// Which source the system clock should be generated from.
+pub enum SystemClockSource {
+    /// Use the internal digital frequency locked loop (DFLL) sourced from
+    /// the internal RC32K clock. Note this typically requires calibration
+    /// of the RC32K to have a consistent clock.
+    DfllRc32k,
+
+    /// Use an external crystal oscillator as the direct source for the
+    /// system clock.
+    ExternalOscillator,
+}
+
+const PM_BASE: usize = 0x400E0000;
+const BSCIF_BASE: usize = 0x400F0400;
+const SCIF_BASE: usize = 0x400E0800;
+const FLASHCALW_BASE: usize = 0x400A0000;
+
+const HSB_MASK_OFFSET: u32 = 0x24;
+const PBA_MASK_OFFSET: u32 = 0x28;
+const PBB_MASK_OFFSET: u32 = 0x2C;
+const PBD_MASK_OFFSET: u32 = 0x34;
+
+static mut PM: *mut PmRegisters = PM_BASE as *mut PmRegisters;
+static mut BSCIF: *mut BscifRegisters = BSCIF_BASE as *mut BscifRegisters;
+static mut SCIF: *mut ScifRegisters = SCIF_BASE as *mut ScifRegisters;
+static mut FLASHCALW: *mut FlashcalwRegisters = FLASHCALW_BASE as *mut FlashcalwRegisters;
+
+static mut SYSTEM_FREQUENCY: u32 = 0;
```

Comment:
I believe this works, but for safety this should be a `VolatileCell` to make sure the compiler doesn't discard reads/writes to it.